### PR TITLE
Migrate correspondence

### PR DESCRIPTION
**Pull request by jebbinSCL**

### CODE REVIEW

File: TrelloMigrate/SrmApiClient.fs [hunk 1]
```diff
@@ -25,4 +25,10 @@ module Session =
...
     let postAndGetId (session : Session) = 
-        { session with Id = postAndGetGuid endpoint session } 
\ No newline at end of file
+        { session with Id = postAndGetGuid endpoint session } 
+
+module Correspondence = 
+    let private endpoint = Uri(commsServiceUri, "Correspondence")
+
+    let post (item : CorrespondenceItem) = 
+        post endpoint item |> ignore
```
Comment (jebbinSCL) [inline TrelloMigrate/SrmApiClient.fs]: Tiny refactoring will be done in next pull request.

File: TrelloMigrate/Transform.fs [hunk 2]
```diff
@@ -118,26 +119,90 @@ module private SessionAndSpeaker =
...
-    //TODO perform a merge of speaker and admin information to make sure no information is lost. Currently extra handles would be dropped. 
-    let sessions, nonAdminSpeakerOptions = 
-        parsedCards
-        |> Array.map (fun ss -> 
-            let foundSpeakerAsAdmin = 
-                admins |> Array.tryFind(fun a -> a.ProfileWithHandles.Profile.Forename = ss.Speaker.Profile.Forename && a.ProfileWithHandles.Profile.Surname = ss.Speaker.Profile.Surname)
-            match foundSpeakerAsAdmin with 
-            | Some adminWithRef -> 
-                {Session = ss.Session; SpeakerTrelloId = adminWithRef.ReferenceId; AdminTrelloId = ss.AdminTrelloId}, None
+    let splitProfilesFromSessions (admins : ProfileWithReferenceId []) (parsedCards : ParsedCard []) =
```
Comment (jebbinSCL) [inline TrelloMigrate/Transform.fs]: Github hasn't dealt with this nicely, but this function hasn't changed, only moved within the SessionAndSpeaker module.

File: TrelloMigrate/Transform.fs [hunk 3]
```diff
@@ -118,26 +119,90 @@ module private SessionAndSpeaker =
...
+        let success, exactDate = 
+            DateTime.TryParseExact(dateString, [| "dd/mm/yy"; "dd/mm/yyyy"; "dd.mm.yy"; "dd.mm.yyyy"; "dd/mm.yyyy"; "dd.mm/yyyy" |], CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal)
+        if success then exactDate
+        else 
+            try
+                DateTime.Parse(dateString, CultureInfo.InvariantCulture)
+            with
+            | ex -> failwith <| sprintf "Parsing date for correspondence action: '%s' failed with exception: %A" actionText ex
+
+    let private selectSenderAndReceiver (actionText : string) (dirString : string) speakerValue adminValue =
```
Comment (pallenSL) [inline TrelloMigrate/Transform.fs]: Is "direction" a keyword?

File: TrelloMigrate/Transform.fs [hunk 4]
```diff
@@ -118,26 +119,90 @@ module private SessionAndSpeaker =
...
+    let private selectSenderAndReceiver (actionText : string) (dirString : string) speakerValue adminValue = 
+        match dirString.ToUpperInvariant() with
+        | "SEND" -> adminValue, speakerValue
+        | "RECEIVE" | "RECIEVE" -> speakerValue, adminValue
+        | _ -> failwith <| sprintf "Parsing directon for correspondence action: '%s' failed" actionText
+
+    let private tryCreateCorrespondence (cardTitle : string) (speakerRefId : string) (speakerEmail : string) (adminRefIdOption : string option) (adminEmailOption : string option) (action : BasicAction) = 
+        let m = Regex.Match(action.Data.Text, "\[(?<date>[0-9\/\.]+) - (?<direction>[a-zA-Z]+)\](?<message>.*)$", RegexOptions.ExplicitCapture ||| RegexOptions.Singleline)
+        if m.Success then 
+            match adminEmailOption, adminRefIdOption with
```
Comment (pallenSL) [inline TrelloMigrate/Transform.fs]: Surely this should be the outer test?

Comment (jebbinSCL) [inline TrelloMigrate/Transform.fs]: Nope. We only need to check if there is an admin if there are correspondence comments on the card. So, if non of the comments contain correspondence, this test won't be hit.

File: TrelloMigrate/Transform.fs (hunk 3 above)
Comment (jebbinSCL) [inline TrelloMigrate/Transform.fs]: I'll change this now.

Comment (jebbinSCL) [inline TrelloMigrate/Transform.fs]: Adjusted.

**Comment by pallenSL** [comment]
:shipit:

**Pull request merged**